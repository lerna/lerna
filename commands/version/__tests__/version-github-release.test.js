"use strict";

// local modules _must_ be explicitly mocked
jest.mock("../lib/git-add");
jest.mock("../lib/git-commit");
jest.mock("../lib/git-push");
jest.mock("../lib/git-tag");
jest.mock("../lib/is-anything-committed");
jest.mock("../lib/is-behind-upstream");
jest.mock("../lib/remote-branch-exists");

// mocked modules
const { client } = require("@lerna/github-client");
const { recommendVersion } = require("@lerna/conventional-commits");

// helpers
const initFixture = require("@lerna-test/init-fixture")(__dirname);

// test command
const lernaVersion = require("@lerna-test/command-runner")(require("../command"));

test("--github-release does not create a release if --no-push is passed", async () => {
  const cwd = await initFixture("independent");

  await lernaVersion(cwd)("--github-release", "--conventional-commits", "--no-push");

  expect(client.repos.createRelease).not.toHaveBeenCalled();
});

test("--github-release throws an error if --conventional-commits is not passed", async () => {
  const cwd = await initFixture("independent");

  try {
    await lernaVersion(cwd)("--github-release");
  } catch (err) {
    expect(err.message).toBe("To create a Github Release, you must enable --conventional-commits");
    expect(client.repos.createRelease).not.toHaveBeenCalled();
  }

  expect.hasAssertions();
});

test("--github-release marks a version as a pre-release if it contains a valid part", async () => {
  const cwd = await initFixture("normal");

  recommendVersion.mockResolvedValueOnce("2.0.0-alpha.1");

  await lernaVersion(cwd)("--github-release", "--conventional-commits");

  expect(client.repos.createRelease).toHaveBeenCalledTimes(1);
  expect(client.repos.createRelease).toHaveBeenCalledWith({
    owner: "lerna",
    repo: "lerna",
    tag_name: "v2.0.0-alpha.1",
    name: "v2.0.0-alpha.1",
    body: "normal",
    draft: false,
    prerelease: true,
  });
});

test("--github-release creates a release for every independent version", async () => {
  const cwd = await initFixture("independent");
  const versionBumps = new Map([
    ["package-1", "1.0.1"],
    ["package-2", "2.1.0"],
    ["package-3", "4.0.0"],
    ["package-4", "4.1.0"],
    ["package-5", "5.0.1"],
  ]);

  versionBumps.forEach(bump => recommendVersion.mockResolvedValueOnce(bump));

  await lernaVersion(cwd)("--github-release", "--conventional-commits");

  expect(client.repos.createRelease).toHaveBeenCalledTimes(5);
  versionBumps.forEach((version, name) => {
    expect(client.repos.createRelease).toHaveBeenCalledWith({
      owner: "lerna",
      repo: "lerna",
      tag_name: `${name}@${version}`,
      name: `${name}@${version}`,
      body: `${name} - ${version}`,
      draft: false,
      prerelease: false,
    });
  });
});

test("--github-release creates a single fixed release", async () => {
  const cwd = await initFixture("normal");

  recommendVersion.mockResolvedValueOnce("1.1.0");

  await lernaVersion(cwd)("--github-release", "--conventional-commits");

  expect(client.repos.createRelease).toHaveBeenCalledTimes(1);
  expect(client.repos.createRelease).toHaveBeenCalledWith({
    owner: "lerna",
    repo: "lerna",
    tag_name: "v1.1.0",
    name: "v1.1.0",
    body: "normal",
    draft: false,
    prerelease: false,
  });
});
