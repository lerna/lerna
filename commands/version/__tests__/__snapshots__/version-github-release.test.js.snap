// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`--github-release fixed should create a single release 1`] = `
"v2.0.0

HEAD -> refs/heads/master, tag: refs/tags/v2.0.0

diff --git a/CHANGELOG.md b/CHANGELOG.md
new file mode 100644
index SHA..SHA
--- /dev/null
+++ b/CHANGELOG.md
@@ -0,0 +1 @@
+changelog
\\\\ No newline at end of file
diff --git a/lerna.json b/lerna.json
index SHA..SHA 100644
--- a/lerna.json
+++ b/lerna.json
@@ -2 +2 @@
-  \\"version\\": \\"1.0.0\\"
+  \\"version\\": \\"2.0.0\\"
diff --git a/packages/package-1/CHANGELOG.md b/packages/package-1/CHANGELOG.md
new file mode 100644
index SHA..SHA
--- /dev/null
+++ b/packages/package-1/CHANGELOG.md
@@ -0,0 +1 @@
+changelog
\\\\ No newline at end of file
diff --git a/packages/package-1/package.json b/packages/package-1/package.json
index SHA..SHA 100644
--- a/packages/package-1/package.json
+++ b/packages/package-1/package.json
@@ -3 +3 @@
-  \\"version\\": \\"1.0.0\\"
+  \\"version\\": \\"2.0.0\\"
diff --git a/packages/package-2/CHANGELOG.md b/packages/package-2/CHANGELOG.md
new file mode 100644
index SHA..SHA
--- /dev/null
+++ b/packages/package-2/CHANGELOG.md
@@ -0,0 +1 @@
+changelog
\\\\ No newline at end of file
diff --git a/packages/package-2/package.json b/packages/package-2/package.json
index SHA..SHA 100644
--- a/packages/package-2/package.json
+++ b/packages/package-2/package.json
@@ -3 +3 @@
-  \\"version\\": \\"1.0.0\\",
+  \\"version\\": \\"2.0.0\\",
@@ -5 +5 @@
-    \\"package-1\\": \\"^1.0.0\\"
+    \\"package-1\\": \\"^2.0.0\\"
diff --git a/packages/package-3/CHANGELOG.md b/packages/package-3/CHANGELOG.md
new file mode 100644
index SHA..SHA
--- /dev/null
+++ b/packages/package-3/CHANGELOG.md
@@ -0,0 +1 @@
+changelog
\\\\ No newline at end of file
diff --git a/packages/package-3/package.json b/packages/package-3/package.json
index SHA..SHA 100644
--- a/packages/package-3/package.json
+++ b/packages/package-3/package.json
@@ -3 +3 @@
-  \\"version\\": \\"1.0.0\\",
+  \\"version\\": \\"2.0.0\\",
@@ -8 +8 @@
-    \\"package-2\\": \\"^1.0.0\\"
+    \\"package-2\\": \\"^2.0.0\\"
diff --git a/packages/package-4/CHANGELOG.md b/packages/package-4/CHANGELOG.md
new file mode 100644
index SHA..SHA
--- /dev/null
+++ b/packages/package-4/CHANGELOG.md
@@ -0,0 +1 @@
+changelog
\\\\ No newline at end of file
diff --git a/packages/package-4/package.json b/packages/package-4/package.json
index SHA..SHA 100644
--- a/packages/package-4/package.json
+++ b/packages/package-4/package.json
@@ -3 +3 @@
-  \\"version\\": \\"1.0.0\\",
+  \\"version\\": \\"2.0.0\\",
diff --git a/packages/package-5/CHANGELOG.md b/packages/package-5/CHANGELOG.md
new file mode 100644
index SHA..SHA
--- /dev/null
+++ b/packages/package-5/CHANGELOG.md
@@ -0,0 +1 @@
+changelog
\\\\ No newline at end of file
diff --git a/packages/package-5/package.json b/packages/package-5/package.json
index SHA..SHA 100644
--- a/packages/package-5/package.json
+++ b/packages/package-5/package.json
@@ -4 +4 @@
-    \\"package-1\\": \\"^1.0.0\\"
+    \\"package-1\\": \\"^2.0.0\\"
@@ -7 +7 @@
-    \\"package-3\\": \\"^1.0.0\\"
+    \\"package-3\\": \\"^2.0.0\\"
@@ -10 +10 @@
-  \\"version\\": \\"1.0.0\\"
+  \\"version\\": \\"2.0.0\\""
`;

exports[`--github-release independent should create a release for every package version 1`] = `
"Publish

 - package-1@1.0.1
 - package-2@2.1.0
 - package-3@4.0.0
 - package-4@4.1.0
 - package-5@5.0.1

HEAD -> refs/heads/master, tag: refs/tags/package-5@5.0.1, tag: refs/tags/package-4@4.1.0, tag: refs/tags/package-3@4.0.0, tag: refs/tags/package-2@2.1.0, tag: refs/tags/package-1@1.0.1

diff --git a/packages/package-1/CHANGELOG.md b/packages/package-1/CHANGELOG.md
new file mode 100644
index SHA..SHA
--- /dev/null
+++ b/packages/package-1/CHANGELOG.md
@@ -0,0 +1 @@
+changelog
\\\\ No newline at end of file
diff --git a/packages/package-1/package.json b/packages/package-1/package.json
index SHA..SHA 100644
--- a/packages/package-1/package.json
+++ b/packages/package-1/package.json
@@ -3 +3 @@
-  \\"version\\": \\"1.0.0\\"
+  \\"version\\": \\"1.0.1\\"
diff --git a/packages/package-2/CHANGELOG.md b/packages/package-2/CHANGELOG.md
new file mode 100644
index SHA..SHA
--- /dev/null
+++ b/packages/package-2/CHANGELOG.md
@@ -0,0 +1 @@
+changelog
\\\\ No newline at end of file
diff --git a/packages/package-2/package.json b/packages/package-2/package.json
index SHA..SHA 100644
--- a/packages/package-2/package.json
+++ b/packages/package-2/package.json
@@ -3 +3 @@
-  \\"version\\": \\"2.0.0\\",
+  \\"version\\": \\"2.1.0\\",
@@ -5 +5 @@
-    \\"package-1\\": \\"^1.0.0\\"
+    \\"package-1\\": \\"^1.0.1\\"
diff --git a/packages/package-3/CHANGELOG.md b/packages/package-3/CHANGELOG.md
new file mode 100644
index SHA..SHA
--- /dev/null
+++ b/packages/package-3/CHANGELOG.md
@@ -0,0 +1 @@
+changelog
\\\\ No newline at end of file
diff --git a/packages/package-3/package.json b/packages/package-3/package.json
index SHA..SHA 100644
--- a/packages/package-3/package.json
+++ b/packages/package-3/package.json
@@ -3 +3 @@
-  \\"version\\": \\"3.0.0\\",
+  \\"version\\": \\"4.0.0\\",
@@ -5 +5 @@
-    \\"package-2\\": \\"^2.0.0\\"
+    \\"package-2\\": \\"^2.1.0\\"
diff --git a/packages/package-4/CHANGELOG.md b/packages/package-4/CHANGELOG.md
new file mode 100644
index SHA..SHA
--- /dev/null
+++ b/packages/package-4/CHANGELOG.md
@@ -0,0 +1 @@
+changelog
\\\\ No newline at end of file
diff --git a/packages/package-4/package.json b/packages/package-4/package.json
index SHA..SHA 100644
--- a/packages/package-4/package.json
+++ b/packages/package-4/package.json
@@ -3 +3 @@
-  \\"version\\": \\"4.0.0\\",
+  \\"version\\": \\"4.1.0\\",
diff --git a/packages/package-5/CHANGELOG.md b/packages/package-5/CHANGELOG.md
new file mode 100644
index SHA..SHA
--- /dev/null
+++ b/packages/package-5/CHANGELOG.md
@@ -0,0 +1 @@
+changelog
\\\\ No newline at end of file
diff --git a/packages/package-5/package.json b/packages/package-5/package.json
index SHA..SHA 100644
--- a/packages/package-5/package.json
+++ b/packages/package-5/package.json
@@ -4 +4 @@
-    \\"package-3\\": \\"^3.0.0\\"
+    \\"package-3\\": \\"^4.0.0\\"
@@ -7 +7 @@
-  \\"version\\": \\"5.0.0\\"
+  \\"version\\": \\"5.0.1\\""
`;
